### PR TITLE
resolve issue #110

### LINKED ISSUE #12: refactor the vshard mode instructions, and settings to indicate to the pool what to use
now this can lead to such unpleasant situations
```
vshardrouter.CallOpts{
    VshardMode: vshardrouter.WriteMode,
    PoolMode:   pool.ANY,
    Timeout:    time.Second * 5,
},
```

### LINKED ISSUE #110: A new interface instead of RouterCallImpl
https://github.com/KaymeKaydex/go-vshard-router/blob/097ac242d44549ce23d0270cfae4dba9862f35b3/api.go#L81-L85

The current interface of the RouterCallImpl method has several drawbacks:
- It performs a redundant job by returning both (see #100)
  - user method's response array as interface{}, which is used rarely
  - StorageResultTypedFunc to unmarshal user's answer
- Allows set both vshard mode and pool mode, but some combinations are incompatible: vshardMode="write" and pool.RO. (see #12)
- StorageResultTypedFunc recodes only the first element of the user method's response array, but it can return up to 3 elements, see https://github.com/tarantool/vshard/blob/dfa2cc8a2aff221d5f421298851a9a229b2e0434/vshard/storage/init.lua#L3130

We should invent a new interface without these drawbacks.

### PR DESCRIPTION
**Pull request by nurzhan-saktaganov**
What has been done? Why? What problem is being solved?

I didn't forget about (remove if it is not applicable):

- [x] Changelog (see [documentation](https://keepachangelog.com/en/1.0.0/) for changelog format)

Related issues:

### CODE REVIEW

**Comment by nurzhan-saktaganov** [comment]
Benchmarks:

```
n.saktaganov@n-saktaganov tnt % make bench
go test -bench=. -benchmem
goos: darwin
goarch: arm64
pkg: github.com/KaymeKaydex/go-vshard-router/tests/tnt
cpu: Apple M3 Pro
BenchmarkCallSimpleInsert_GO_RouterCall-12         10719            109068 ns/op            1937 B/op         37 allocs/op
BenchmarkCallSimpleInsert_GO_Call-12                9900            109099 ns/op            1833 B/op         33 allocs/op
BenchmarkCallSimpleInsert_Lua-12                    7352            164535 ns/op            1228 B/op         20 allocs/op
BenchmarkCallSimpleSelect_GO_RouterCall-12         13227             88042 ns/op            2687 B/op         52 allocs/op
BenchmarkCallSimpleSelect_GO_Call-12               13088             87874 ns/op            2131 B/op         41 allocs/op
BenchmarkCallSimpleSelect_Lua-12                    8695            137136 ns/op            1305 B/op         22 allocs/op
PASS
ok      github.com/KaymeKaydex/go-vshard-router/tests/tnt       21.921s
```

File: api.go [hunk 1]
```diff
@@ -312,30 +369,101 @@ func (r *Router) RouterCallImpl(ctx context.Context,
...
+	fnc string, args interface{}, opts VshardCallOptions) (VshardCallResp, error) {
+	return r.Call(ctx, bucketID, VshardCallModeBRE, fnc, args, opts)
+}
+
+// RouterCallImpl Perform shard operation function will restart operation
+// after wrong bucket response until timeout is reached
+// Deprecated: RouterCallImpl is deprecated.
+// See https://github.com/KaymeKaydex/go-vshard-router/issues/110.
+// Use Call, Call[RO, RW, RE, BRO, BRE] methods instead.
+func (r *Router) RouterCallImpl(ctx context.Context,
```
Comment (nurzhan-saktaganov) [inline api.go:420]: +get rid of RouterCallImpl in `func (r *Router) Do(req *CallRequest, userMode pool.Mode) *CallResponse {`

File: api.go [hunk 2]
```diff
@@ -176,54 +178,111 @@ type CallOpts struct {
...
+type VshardCallMode int
 
-// RouterCallImpl Perform shard operation function will restart operation
-// after wrong bucket response until timeout is reached
-func (r *Router) RouterCallImpl(ctx context.Context,
-	bucketID uint64,
-	opts CallOpts,
-	fnc string,
-	args interface{}) (interface{}, StorageResultTypedFunc, error) {
+const (
```
Comment (nurzhan-saktaganov) [inline api.go:183]: Add comments for public/exported stuff

**Comment by KaymeKaydex** [comment]
looks very cool, as discussed we will fix the standard implementation of RouterCallImpl, although it is not optimal, since it decodes empty interfaces. Standard Call will allow to fix this. And also the next pr we will fix the sugar resolution and make it optimal

File: api.go (hunk 1 above)
Comment (nurzhan-saktaganov) [reply api.go]: #120

File: api.go [hunk 3]
```diff
@@ -167,62 +169,180 @@ func (s StorageCallVShardError) Error() string {
...
+	}
+
+	return nil
+}
 
 // RouterCallImpl Perform shard operation function will restart operation
 // after wrong bucket response until timeout is reached
+// Deprecated: RouterCallImpl is deprecated.
+// See https://github.com/KaymeKaydex/go-vshard-router/issues/110.
+// Use Call, Call[RO, RW, RE, BRO, BRE] methods instead.
```
Comment (KaymeKaydex) [inline api.go:235]:
```suggestion
// Use Call, Call with RO, RW, RE, BRO, BRE methods instead.
```
cause call is non generic

File: replicaset.go [hunk 4]
```diff
@@ -120,16 +120,19 @@ func bucketStatWait(future *tarantool.Future) (BucketStatInfo, error) {
 }
 
 // ReplicaCall perform function on remote storage
-// link https://github.com/tarantool/vshard/blob/master/vshard/replicaset.lua#L661
-// This method is deprecated, because looks like it has a little bit broken interface
+// link https://github.com/tarantool/vshard/blob/99ceaee014ea3a67424c2026545838e08d69b90c/vshard/replicaset.lua#L661
+// Deprecated: RouterCallImpl is deprecated,
```
Comment (KaymeKaydex) [inline replicaset.go:124]:
```suggestion
```
method comment mismatch

**Review by KaymeKaydex — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 5]
```diff
@@ -19,6 +19,7 @@ FEATURES:
 * Add pause between requests in buckets discovering. Configured by config DiscoveryWorkStep, default is 10ms.
 * Add ReplicaUUID to the StorageCallVShardError struct.
 * New method 'RouterMapCallRW[T]' to replace the deprecated one 'RouterMapCallRWImpl'.
+* New methods 'Router.Call[XYZ]' to replace the deprecated one 'RouterCallImpl'.
```
[the remark this answers is not in the document]
Comment (nurzhan-saktaganov) [reply CHANGELOG.md]: fixed

File: api.go (hunk 3 above)
Comment (nurzhan-saktaganov) [reply api.go]: fixed

File: replicaset.go (hunk 4 above)
Comment (nurzhan-saktaganov) [reply replicaset.go]: fixed

File: api.go [hunk 6]
```diff
@@ -22,6 +22,9 @@ type VshardMode string
 const (
 	ReadMode  VshardMode = "read"
 	WriteMode VshardMode = "write"
+
+	// callTimeoutDefault is a default timeout when no timeout is provided
+	callTimeoutDefault = 500 * time.Millisecond
```
[the remark this answers is not in the document]
Comment (nurzhan-saktaganov) [reply api.go:27]: Since this value is intended to be used only for internal purposes, I guess we should leave it non-public.

**Review by KaymeKaydex — APPROVED** [review]

**Pull request merged**